RecoEgamma/EgammaTools/src/HGCalShowerShapeHelper.cc

```diff
@@ -0,0 +1,310 @@
+                                           double minHitET,
+                                           int minLayer,
+                                           int maxLayer,
+                                           DetId::Detector subDet) {
+  // Safety checks
+  nLayer_ = maxLayer - minLayer + 1;
+  assert(nLayer_ > 0);
+
+  minHitE_ = minHitE;
+  minHitET_ = minHitET;
+  minHitET2_ = minHitET * minHitET;
+  minLayer_ = minLayer;
+  maxLayer_ = maxLayer;
+  subDet_ = subDet;
+
+  setFilteredHitsAndFractions(hitsAndFracs);
+
+  setLayerWiseStuff();
+}
+
+void HGCalShowerShapeHelper::setPFRecHitPtrMap(const std::vector<reco::PFRecHit> &recHits) {
+  pfRecHitPtrMap_.clear();
+
+  for (auto &recHit : recHits) {
+    pfRecHitPtrMap_[recHit.detId()] = &recHit;
+  }
+}
+
+void HGCalShowerShapeHelper::setFilteredHitsAndFractions(const std::vector<std::pair<DetId, float> > &hitsAndFracs) {
+  hitsAndFracs_.clear();
+  hitEnergies_.clear();
+  hitEnergiesWithFracs_.clear();
+
+  for (const auto &hnf : hitsAndFracs) {
+    DetId hitId = hnf.first;
+    float hitEfrac = hnf.second;
+
+    int hitLayer = recHitTools_.getLayer(hitId);
+
+    if (hitLayer > nLayer_) {
+      continue;
+    }
+
+    if (hitId.det() != subDet_) {
+      continue;
+    }
+
+    if (pfRecHitPtrMap_.find(hitId.rawId()) == pfRecHitPtrMap_.end()) {
+      continue;
+    }
+
+    reco::PFRecHit recHit = *pfRecHitPtrMap_[hitId.rawId()];
+
+    if (recHit.energy() < minHitE_) {
+      continue;
+    }
+
+    if (recHit.pt2() < minHitET2_) {
+      continue;
+    }
+
+    // Fill the vectors
+    hitsAndFracs_.push_back(hnf);
+    hitEnergies_.push_back(recHit.energy());
+    hitEnergiesWithFracs_.push_back(recHit.energy() * hitEfrac);
+  }
+}
+
+void HGCalShowerShapeHelper::setLayerWiseStuff() {
+  layerEnergies_.clear();
+  layerEnergies_.resize(nLayer_);
+
+  layerCentroids_.clear();
+  layerCentroids_.resize(nLayer_);
+
+  centroid_.SetXYZ(0, 0, 0);
+
+  int iHit = -1;
+  double totalW = 0;
+
+  // Compute the centroid per layer
+  for (const auto &hnf : hitsAndFracs_) {
+    iHit++;
+
+    DetId hitId = hnf.first;
+
+    double weight = hitEnergies_[iHit];
+    totalW += weight;
+
+    const auto &hitPos = recHitTools_.getPosition(hitId);
+    ROOT::Math::XYZVector hitXYZ(hitPos.x(), hitPos.y(), hitPos.z());
+
+    centroid_ += weight * hitXYZ;
+
+    int hitLayer = recHitTools_.getLayer(hitId) - 1;
+
+    layerEnergies_[hitLayer] += weight;
+    layerCentroids_[hitLayer] += weight * hitXYZ;
+  }
+
+  int iLayer = -1;
+
+  for (auto &centroid : layerCentroids_) {
+    iLayer++;
+
+    if (layerEnergies_[iLayer]) {
+      centroid /= layerEnergies_[iLayer];
+    }
+  }
+
+  if (totalW) {
+    centroid_ /= totalW;
+  }
+}
+
+double HGCalShowerShapeHelper::getCellSize(DetId detId) {
+  double siThickness = recHitTools_.getSiThickness(detId);
+
+  // HD wafers
+  if (siThickness < 150) {
+    return 0.465;
+  }
+
+  // LD wafers
+  else {
```

Comment:
suprised this wasnt caught by code checks, it would be good to have the else more linked to the if, theres too much space between them, it they unconnected

Reply:
personally I would have done  return siThickness < 150 ? kSmallCellSize : kLargeCellSize;

and defined kSmallCellSize in the class declaration as a const static double (its good to not have magic numbers buried in the code). Feel free to change the kSmallCellSize name to something else more appropriate (I use k to indicate constant although its personal style and not part of the rules.

Reply:
ah perhaps kHDWaferCellSize, kLDWaferCellSize